Add support for FLPR core in nRF54H20

gmarull opened the pull request:
This PR allows building for the FLPR (Fast Lightweight Processor) present in nRF54H20.

gmarull committed:
snippets: nordic-flpr: move vevif enablement to nRF54L15DK overlay

Other platforms using FLPR, e.g. nRF54H20 do not need this.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
snippets: nordic-flpr: add support for nrf54h20dk

Provide overlay to mark uart135 as reserved (used by FLPR).

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
modules: hal_nordic: add support for nRF54H20 FLPR

Add necessary HAL glue code to support nRF54H20.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
dts: nordic: nrf54h20: define cpuflpr

Define the FLPR VPR CPU instance.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
dts: nordic: nrf54h20: define cpuflpr_vpr coprocessor

Add a new entry for the FLPR co-processor instance.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
dts: nordic: nrf54h20: define FLPR CLIC instance

Define the FLPR CLIC instance.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
dts: nordic: nrf54h20: define cpuflpr VEVIF RX instance

Define the FLPR VEVIF instance (used to receive _tasks_).

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
dts: nordic: nrf54h20: define cpuflpr VEVIF TX instance

Define the FLPR VEVIF instance (used to send _tasks_).

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
dts: riscv: nordic: nrf54h20: introduce cpuflpr

Add a new base devicetree file for the FLPR core.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
soc: nordic: nrf54h20: define CPUFLPR core

So that FLPR core can be used.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
soc: nordic: vpr: allow building VPR launcher for FLPR

VPR launcher can also be used for FLPR.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
boards: nrf54h20dk: add cpuflpr target

Add a new board target to build for the FLPR core.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
boards: nrf54h20dk: push dma_fast_region area to the end of the region

So that we can use the first 48K for FLPR.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
boards: nrf54h20dk: shrink cpuapp slot0 partition

So that we have an extra 48K for the FLPR core.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
boards: nrf54h20: define FLPR memory layout within RAM21 region

Define the FLPR memory layout:

- 46K for code data
- 1K/1K for IPC between cpuapp/cpuflpr

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
boards: nrf54h20dk: add FLPR code partition in MRAM1X

Add a new 48K partition for FLPR in the MRAM1X region.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
samples: sysbuild: hello_world: add cpuapp/cpuflpr configuration

So that a hello world that runs on cpuapp/cpuflpr cores can be easily
tested.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
boards: nrf54h20dk: add cpuapp/cpuflpr IPC configuration

Add IPC configuration to allow communicating between cpuapp and cpuflpr
cores.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
snippets: nordic-flpr-xip: move vevif enablement to nRF54L15DK overlay

Other platforms, e.g. nRF54H20 do not need this.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
snippets: nordic-flpr-xip: add support for nrf54h20dk

Add support for nRF54H20.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
boards: nrf54h20dk: add xip variant for cpuflpr

While executing from MRAM will likely never happen in practice for the
FLPR core, add such variant for testing.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
samples: sysbuild: hello_world: add cpuapp/cpuflpr+xip configuration

So that xip board variant can be easily tested.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
dts: nordic: nrf54h20: fix uarte120 IRQ

It's 230, not 229.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
dts: bindings: clock: fixed-clock: include base.yaml

Include base.yaml, so that properties like `clocks` can be optionally
set.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
dts: nordic: nrf54h20: define hsfll120 instance

HSFLL120 is a clock used by fast peripherals, and it is controlled by
the system controller. From an application perspective, it is a fixed
clock. Note that it has multiple outputs (clk_main @ 320MHz, and
clk_main2|4 which provide the main clock divided by 2 and 4,
respectivelu). Only the main frequency is represented for now.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
dts: nordic: nrf54h20: set clocks for uart120 instance

UARTE120 is clocked by HFSFLL120 instance.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
drivers: serial: nrfx_uarte: add support for arbitrary peripheral clock

Some UARTE instances may be clocked at higher speeds than 16MHz, so the
baudrate setting needs to be scaled accordingly. This patch parses the
`clocks` property and obtains the clock-frequency property of the
associated clock, assuming it is a fixed-clock. We should ideally have a
proper clock control subsystem where frequency can be queried using an
API, but we're far from there.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
dts: bindings: pinctrl: nrf: add nordic,clock-enable

On new SoCs, certain pins need to enable the clock setting on the pin
for it to work properly. For now, this has been handled internally in
the pinctrl driver, however, it appears to be an instance-specific
property (e.g. UARTE/SPIM instances in the fast domain do not require
such setting). Move the configuration of this settings to DT, the best
place where to have instance-specific hardware settings.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
drivers: pinctrl: nrf: add support for nordic,clock-enable

Driver will be capable of retrieving such property from DT and apply it
accordingly.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
boards, tests: nrf54h20: add nordic,clock-enable property

Applies to:

- UART/E (TXD)
- SPIM (SCK, MOSI)
- SPIS (SCK, MISO)
- TWIM (SDA, SCL)

Except fast-instances.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

hubertmis commented:
> * VEVIF/BELLBOARD IPC channels for cpuapp<>cpuflpr?

I've created https://github.com/nrfconnect/doc-internal/pull/112 to specify this.

hubertmis commented on dts/riscv/nordic/nrf54h20_cpuflpr.dtsi:
```diff
@@ -0,0 +1,51 @@
+	status = "okay";
+};
+
+&cpusec_bellboard {
+	compatible = "nordic,nrf-bellboard-tx";
+};
```
This is not required. FLPR shall not communicate directly with SecDom

gmarull replied:
done

hubertmis commented on dts/riscv/nordic/nrf54h20_cpuflpr.dtsi:
```diff
@@ -0,0 +1,51 @@
+	compatible = "nordic,nrf-bellboard-tx";
+};
+
+&cpurad_bellboard {
+	compatible = "nordic,nrf-bellboard-tx";
+};
```
This also should be removed. FLPR should not communicate with Radio. Only with App.

gmarull replied:
done

hubertmis commented on boards/nordic/nrf54h20dk/nrf54h20dk_nrf54h20-memory_map.dtsi:
```diff
@@ -100,10 +100,23 @@
 			perm-read;
 			perm-write;
 			perm-secure;
+			perm-execute;
```
I'm thinking if we should create another `"nordic,owned-memory"` region with RWSX permission that would include `cpuflpr_code_data` and keep this one as RWS. We should have enough hardware resources to do that. It would improve security a little.

hubertmis replied:
Oh, I see I caused unnecessary actions here. I forgot FLPR has a direct bridge to RAM21 bypassing MPC. So it does require `perm-execute` to execute from RAM21. All regions can be inside a single `cpuapp_ram21_region` with permissions RWS.

I think it would be good to add a comment by `cpuflpr_code_data` to clarify why it is in a region without `perm-execute` because, as we can see in this review, it is counter-intuitive.

gmarull replied:
fixed, added note

hubertmis commented on boards/nordic/nrf54h20dk/nrf54h20dk_nrf54h20-ipc_conf.dtsi:
```diff
@@ -41,6 +41,13 @@
+			compatible = "zephyr,ipc-icmsg";
+			status = "disabled";
+			mboxes = <&cpuapp_bellboard 14>,
+				 <&cpuflpr_vevif 13>;
```
This one should be "12" to be consistent with the design in https://github.com/nrfconnect/doc-internal/pull/112

gmarull replied:
Note: I see that for cpuflpr vevif, the tasks mask is 0xffff0000, so if I got this correctly, only 31-16 can be used (driver is designed to handle this way right now).

hubertmis replied:
Do you mean VEVIF task 12 does not work? If that's the case, the channels allocation should be redesigned and doc-internal updated to reflect this

gmarull replied:
![image](https://github.com/zephyrproject-rtos/zephyr/assets/25011557/c0759624-ac41-4974-9740-aec11612abc4)

It doesn't seem to work (just tested quickly)

hubertmis replied:
Could you try `<&cpuflpr_vevif 16>;`? If it works as expected I'll update the docs accordingly.

gmarull replied:
yes, 16 works fine

hubertmis commented on snippets/nordic-flpr/nordic-flpr.overlay:
```diff
@@ -6,7 +6,3 @@
 	status = "okay";
 };
-
-&cpuapp_vevif_tx {
```
This could be useful for more complex IPC schemes with more channels required. But I guess in such a scenario, it could be added back as a snippet or an overlay.

gmarull replied:
The snipper just covers remote core boot-up, not any extra things. I think it should be a separate snippet.

hubertmis reviewed (changes requested):
The memory layout looks fine to me.

Also, it would be nice to have an IPC test target for app-flpr communication.

fabiobaltieri commented on dts/riscv/nordic/nrf54h20_cpuflpr.dtsi:
```diff
@@ -0,0 +1,51 @@
+cpuflpr_vevif: &cpuflpr_vevif_rx {};
+cpusys_vevif: &cpusys_vevif_tx {};
+
+/delete-node/ &cpuapp;
+/delete-node/ &cpuapp_peripherals;
+/delete-node/ &cpuapp_ppb;
+/delete-node/ &cpuapp_ram0;
+/delete-node/ &cpurad;
+/delete-node/ &cpurad_peripherals;
+/delete-node/ &cpurad_ppb;
+/delete-node/ &cpurad_ram0;
+/delete-node/ &cpuppr;
```
Wouldn't you be better off reworking `nordic/nrf54h20.dtsi` into multiple files? like a "common" one and others for the specific domain. It's a bit odd to have so many delete-node

gmarull replied:
Base file has been designed as a pseudo-system-devicetree.

gmarull committed:
samples: drivers: mbox: add support for nrf54h20dk cpuapp<>cpuflpr

Add support for testing IPC between cpuapp<>cpuflpr. Some adjustments
have been done to be able to support two different remote targets for
the nrf54h20dk cpuapp.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull commented:
> The memory layout looks fine to me.
> 
> Also, it would be nice to have an IPC test target for app-flpr communication.

Added support for `samples/drivers/mbox`

masz-nordic commented on dts/common/nordic/nrf54h20.dtsi:
```diff
@@ -429,6 +486,34 @@
+					nordic,tasks-mask = <0xffff0000>;
+				};
+
+				cpuflpr_clic: interrupt-controller@1000 {
```
CLIC should be at `0xF0000000` for all VPRs, like in L15:
https://github.com/zephyrproject-rtos/zephyr/blob/main/dts/common/nordic/nrf54l15.dtsi#L148
Then assertion in `soc/nordic/validate_base_addresses.c` should pass.

gmarull replied:
IIRC all CLICs are visible from eg app core, so they can't share the same address. Maybe `0xF0000000` is the address as seen from each VPR instance? @hubertmis

hubertmis replied:
It seems like there is a misunderstanding here. Looking at the Product Specification documentation, I see that the CLIC registers can be instantiated at different addresses for each VPR instance. In nRF54H20, all of them are at 0xf0000000. This address is available only from a VPR CPU. I don't see any way for a VPR's owner to access CLIC registers.

The address 0x00001000 is a default address for a CLIC instance (default MCLICBASE value), but it is overwritten for each VPR instance in the system.

hubertmis replied:
And I can't find anything at address 0x8d5000 (0x8d4000 + 0x1000 offset)

gmarull commented on dts/bindings/pinctrl/nordic,nrf-pinctrl.yaml:
```diff
@@ -119,3 +119,9 @@ child-binding:
         description: |
           Invert pin polarity (set the active state to low).
           Only valid for PWM channel output pins.
+
+      nordic,clock-enable:
+        type: boolean
+        description: |
+          Enable clock for the pin. Only valid or required for certain
+          peripherals, refer to the reference manual.
```
I understand your concerns, you're totally right. But I guess it's similar if you e.g. forget to configure I2C pins in open-drain mode, or that you need to specify the "high-drive" mode when operating at certain speeds. New devices will actually become more complicated, as the "route wherever you want" does not apply anymore (the low speed peripheral world is over). I wish a more standard pinmux technique had been implemented, this clock setting is confusing. This setting will be documented in the reference manual somehow, so users will need to understand this when setting up their board pinctrl. Note that this PR changes a few overlays because there are no "default" SPI pins on the DK (not sure why), whereas this does not happen for e.g. uart, can, etc.

bjarki-andreasen replied:
We could add predefined autogenerated pinctrl groups for the peripherals which can only be routed to fixed IOs, similar to what `st` does with their hal https://github.com/zephyrproject-rtos/hal_stm32/blob/main/dts/st/h5/stm32h503cbtx-pinctrl.dtsi , we can add these fun extra configurations like `nordic,clock-enable` to those like:

```
/omit-if-no-ref/ spi130_default: spi130_default {
	group1 {
		psels = <NRF_PSEL(SPIM_SCK, 0, 0)>,
			<NRF_PSEL(SPIM_MISO, 0, 6)>,
			<NRF_PSEL(SPIM_MOSI, 0, 7)>;
		nordic,clock-enable;
	};

	group2 {
		psels = <NRF_PSEL(SPIM_MISO, 0, 6)>;
	};
};
```
and have the user simply add
```
&spi130 {
	pinctrl-0 = <&spi130_default_alt>;
	pinctrl-1 = <&spi130_sleep_alt>;
	pinctrl-names = "default", "sleep";
	...
};
```
Not sure SPI is the best example, TWI is probably better :)

bjarki-andreasen replied:
I will say that it is way beyond the scope of this PR though, I see no better simple alternative to the dts flag :)

nordic-krch replied:
> it's similar if you e.g. forget to configure I2C pins in open-drain mode

I don't think so. open-drain is something that i2c user may expect and clock-enable is something totally unexpected to the user. It's difficult to explain to the user why it is needed and even worst which instance needs it. Imo this should be property of a node (e.g. `uart136 { clock-pin; }` in nrf54h20.dtsi). Though, it might be hard to transfer that from node to pinctrl. I only came with an idea where during the initialization you create pinctrl config copy on stack and patch the configuration. Not sure if that is violating pinctrl api.

gmarull replied:
What about cases where a certain drive mode is needed to operate at high speed? You also need the reference manual to know this is required. I agree it's not ideal, but this setting being instance dependent really made me think it's the "less bad solution". This may be improved in future SoCs to make software easier @hubertmis

gmarull replied:
Another thing I've found: on I3C, unless you configure the `nordic,drive-mode = <NRF_DRIVE_H0D1>;`, waveforms do not keep up (due to its _high-speed_ I guess).

hubertmis commented:
In nRF54L, a similar feature (VPR support) is introduced as a snippet: https://github.com/zephyrproject-rtos/zephyr/blob/main/snippets/nordic-flpr/boards/nrf54l15pdk_nrf54l15_cpuapp.overlay

Should we unify to have consistent solutions across the SoCs?

gmarull commented:
> In nRF54L, a similar feature (VPR support) is introduced as a snippet: https://github.com/zephyrproject-rtos/zephyr/blob/main/snippets/nordic-flpr/boards/nrf54l15pdk_nrf54l15_cpuapp.overlay
> 
> Should we unify to have consistent solutions across the SoCs?

You mean memory layout setup? Because nordic-flpr snippet is also used by 54H20 (see commit list)

hubertmis commented:
> You mean memory layout setup?

Yes, I meant adjusting the memory layout in the snippet - the same way nRF54L does now.

gmarull commented:
> > You mean memory layout setup?
> 
> Yes, I meant adjusting the memory layout in the snippet - the same way nRF54L does now.

We just need to be aware that 54h20 shares certain memory regions for other purposes, so not all config can easily go into a snippet. RAM21 is a good example.

hubertmis commented:
> > > You mean memory layout setup?
> > 
> > 
> > Yes, I meant adjusting the memory layout in the snippet - the same way nRF54L does now.
> 
> We just need to be aware that 54h20 shares certain memory regions for other purposes, so not all config can easily go into a snippet. RAM21 is a good example.

Sure, but the default could be to use the whole RAM21 for the App's DMA needs. The snippet could shrink the App's DMA partition to the last 16 kB and reserve the rest for FLPR. This would be more aligned with the nRF54L approach.

But maybe we should leave it as it is for now and have a more detailed discussion about snippets usage for nRF54H.

nordic-krch commented:
This will require `nrf-regtool` update as it is using nrf-pinctrl layout which changed here. @jonathannilsen @57300

gmarull commented:
> This will require `nrf-regtool` update as it is using nrf-pinctrl layout which changed here. @jonathannilsen @57300

I'll prepare a PR in nrg-regtool when we open the merge window again.

gmarull commented:
rebased, pending https://github.com/nrfconnect/nrf-regtool/pull/148

masz-nordic commented on dts/common/nordic/nrf54h20.dtsi:
```diff
@@ -429,6 +486,34 @@
+					nordic,tasks-mask = <0xffff0000>;
+				};
+
+				cpuflpr_clic: interrupt-controller@1000 {
```
This value is still incorrect.

gmarull replied:
@57300 we need to solve this, CLICs need app level visibility due to e.g. https://github.com/zephyrproject-rtos/zephyr/pull/75451/commits/8b1be5057897bb33b52cc3dee597ab7813884201, so that regtool can route IRQs to the right core, if all have the same address, how will things work?

gmarull replied:
`@0` is not address-less :-)

hubertmis replied:
Right :-)
I meant no bus address meaning no `reg` property. Similar to `cpu` nodes. We could do `cpufpr_clic: interrupt-controller {` as well. I think it would be compliant with dt spec. But let's start by figuring out if CLIC is available from other CPUs or not. Waiting for @57300 's answer in similar question here: https://github.com/zephyrproject-rtos/zephyr/pull/76641#discussion_r1703841262

57300 replied:
The value was incorrect for PPR as well. I proposed a solution in #76691.

gmarull replied:
Implemented @57300 solution here as well.

gmarull replied:
so.... should I go back to initial impl? This means we can't check FLPR address against MDK.

hubertmis replied:
Ok, wait a sec. Now I'm cross-checking and I was mistaken. `@1000` is correct for nRF9230. For nRF54H20 0xF0000000 is the correct address, meaning this is unavailable from cpuapp. So @57300 seems fine

gmarull committed:
modules: hal_nordic: require nrf-regtool 5.6.0

This is required since NRF_FUN_POS (pinctrl) has changed its position
from 17 to 18 (a value hardcoded in the nrf-regtool as well).

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull commented:
Adding DNM, pending nrf-regtool 5.6.0 release (referenced in this PR)

gmarull commented:
@masz-nordic ping

57300 commented on boards/nordic/nrf54h20dk/nrf54h20dk_nrf54h20_cpuflpr.dts:
```diff
@@ -0,0 +1,53 @@
+	};
+};
+
+&cpuapp_cpuflpr_ipc {
+	mbox-names = "tx", "rx";
+	tx-region = <&cpuflpr_cpuapp_ipc_shm>;
+	rx-region = <&cpuapp_cpuflpr_ipc_shm>;
+};
```
Consider adding the inverse of this to `nrf54h20dk_nrf54h20_cpuapp.dts` as well.

gmarull replied:
thx, added

57300 reviewed (commented):
LGTM. One of the commits has a TODO in it, which I think can be dropped.

masz-nordic commented on dts/common/nordic/nrf54h20.dtsi:
```diff
@@ -68,6 +68,56 @@
+				compatible = "nordic,nrf-vevif-task-rx";
+				status = "disabled";
+				interrupt-parent = <&cpuflpr_clic>;
+				interrupts = <0 NRF_DEFAULT_IRQ_PRIORITY>,
+					     <1 NRF_DEFAULT_IRQ_PRIORITY>,
+					     <2 NRF_DEFAULT_IRQ_PRIORITY>,
+					     <3 NRF_DEFAULT_IRQ_PRIORITY>,
+					     <4 NRF_DEFAULT_IRQ_PRIORITY>,
+					     <5 NRF_DEFAULT_IRQ_PRIORITY>,
+					     <6 NRF_DEFAULT_IRQ_PRIORITY>,
+					     <7 NRF_DEFAULT_IRQ_PRIORITY>,
+					     <8 NRF_DEFAULT_IRQ_PRIORITY>,
+					     <9 NRF_DEFAULT_IRQ_PRIORITY>,
+					     <10 NRF_DEFAULT_IRQ_PRIORITY>,
+					     <11 NRF_DEFAULT_IRQ_PRIORITY>,
+					     <12 NRF_DEFAULT_IRQ_PRIORITY>,
+					     <13 NRF_DEFAULT_IRQ_PRIORITY>,
+					     <14 NRF_DEFAULT_IRQ_PRIORITY>,
+					     <15 NRF_DEFAULT_IRQ_PRIORITY>,
```
I think those should be removed since tasks mask is `0xffff0000`. But this can be done in a separate PR.

gmarull replied:
I should check if the code is index-sensitive, if it is, we'd need to change more stuff (like eg use interrupt-names).

masz-nordic replied:
I think it should work, L15 has similar case:
https://github.com/zephyrproject-rtos/zephyr/blob/main/dts/riscv/nordic/nrf54l15_cpuflpr.dtsi#L26-L40

masz-nordic commented on dts/common/nordic/nrf54h20.dtsi:
```diff
@@ -68,6 +68,56 @@
+					     <30 NRF_DEFAULT_IRQ_PRIORITY>,
+					     <31 NRF_DEFAULT_IRQ_PRIORITY>;
+				#mbox-cells = <1>;
+				nordic,tasks = <32>;
```
This should be 16.

gmarull replied:
We need to review all this, I agree it's confusing to have some numbers that are later not allowed by the mask. We need some guidance from @hubertmis here (docs are not so clear).

gmarull replied:
ref 

![image](https://github.com/user-attachments/assets/958b5a34-ac5a-437c-b328-ad47dc7930e3)

hubertmis replied:
I'm too late with the guidance I received from VPR experts. It seems "VEVIF tasks:" should be "16..31", what would translate `<nordic,tasks = <16>;`

gmarull replied:
just checked: code in tasks drivers is index sensitive, as channel number is given in the 0...31 range and it uses this index to pick from an indexed-array. So for now, I'd live driver and DT as is, from a user perspective, nothing changes really. To improve this, we'd need additional information like, do we expect non-contiguous holes in the mask?

hubertmis replied:
> we'd need additional information like, do we expect non-contiguous holes in the mask?

I believe so. I think I've seen VPR instances with non-contiguous events or tasks. Those could have been experimental though. I don't remember the details.

Anyway, I agree there is not rush in improving this.

masz-nordic commented on dts/common/nordic/nrf54h20.dtsi:
```diff
@@ -429,6 +486,24 @@
+					reg = <0x0 0x1000>;
+					status = "disabled";
+					#mbox-cells = <1>;
+					nordic,tasks = <32>;
```
Should be 16.
